charms/reactive/patterns/request_response.py

```diff
@@ -128,6 +128,20 @@ def _load(cls, request_sources):
                 request.response = cls.RESPONSE_CLASS._load(request)
                 cls._cache[request.request_id] = request
 
+    @classmethod
+    def _load_app(cls, relations):
+        for source in relations:
+            if source.application_name is None:
+                continue
+            app_requests = source.to_publish_app
```

Comment:
This is going to fail with a `permission denied` error from Juju on non-leader units.  I ran into the same issue with the newer `loadbalancer` interface, and [worked around it](https://github.com/juju-solutions/loadbalancer-interface/pull/13) by skipping reading the requests and reading the responses directly, but I'm not entirely sure how that would look in this implementation.  I guess `RequesterEndpoint.responses` would need to load the (app-level) responses directly rather than trying to load the (app-level) requests first.

Reply:
Thinking about it more, while this is a problem in general, it shouldn't be a concern for the `prometheus-manual` interface, since only the leader should care if the request failed or not (for the `loadbalancer` interface, the non-leaders still need to see the returned address, but that's not the case for Prometheus).

Reply:
In my case I wasn't able to see any permission denied error on non leader unit when leader-change and manual-job is triggered.